cluster-operator/src/main/java/io/strimzi/operator/cluster/PlatformFeaturesAvailability.java

```diff
@@ -4,23 +4,138 @@
+        Future<PlatformFeaturesAvailability> pfaFuture = Future.future();
+        Future<Void> composeFuture = Future.future();
+        OkHttpClient httpClient = getOkHttpClient(client);
+
+        PlatformFeaturesAvailability pfa = new PlatformFeaturesAvailability();
+
+        Future<VersionInfo> futureVersion = getVersionInfo(vertx, client);
+
+        futureVersion.compose(versionInfo -> {
+            String major = versionInfo.getMajor().equals("") ? Integer.toString(KubernetesVersion.MINIMAL_SUPPORTED_MAJOR) : versionInfo.getMajor();
+            String minor = versionInfo.getMinor().equals("") ? Integer.toString(KubernetesVersion.MINIMAL_SUPPORTED_MINOR) : versionInfo.getMinor();
+            pfa.setKubernetesVersion(new KubernetesVersion(Integer.parseInt(major.split("\\D")[0]), Integer.parseInt(minor.split("\\D")[0])));
+
+            return checkApiAvailability(vertx, httpClient, client.getMasterUrl().toString(), "route.openshift.io", "v1");
+        }).compose(supported -> {
+            pfa.setRoutes(supported);
+            return checkApiAvailability(vertx, httpClient, client.getMasterUrl().toString(), "build.openshift.io", "v1");
+        }).compose(supported -> {
+            pfa.setBuilds(supported);
+            return checkApiAvailability(vertx, httpClient, client.getMasterUrl().toString(), "apps.openshift.io", "v1");
+        }).compose(supported -> {
+            pfa.setApps(supported);
+            return checkApiAvailability(vertx, httpClient, client.getMasterUrl().toString(), "image.openshift.io", "v1");
+        }).compose(supported -> {
+            pfa.setImages(supported);
+            composeFuture.complete();
+        }, composeFuture);
+
+        composeFuture.setHandler(res -> {
+            if (res.succeeded())  {
+                pfaFuture.complete(pfa);
+            } else {
+                pfaFuture.fail(res.cause());
+            }
+        });
+
+        return pfaFuture;
+    }
+
+    private static OkHttpClient getOkHttpClient(KubernetesClient client)   {
+        if (client.isAdaptable(OkHttpClient.class)) {
+            return client.adapt(OkHttpClient.class);
+        } else {
+            log.error("Cannot adapt KubernetesClient to OkHttpClient");
+            throw new RuntimeException("Cannot adapt KubernetesClient to OkHttpClient");
+        }
+    }
+
+    private static Future<VersionInfo> getVersionInfo(Vertx vertx, KubernetesClient client)   {
+        Future<VersionInfo> fut = Future.future();
+
+        vertx.executeBlocking(request -> {
+            try {
+                request.complete(client.getVersion());
+            } catch (Exception e) {
+                log.error("Detection of Kuberetes version failed.", e);
+                request.fail(e);
+            }
+        }, fut.completer());
+
+        return fut;
+    }
+
+    private static Future<Boolean> checkApiAvailability(Vertx vertx, OkHttpClient httpClient, String masterUrl, String api, String version)   {
```

Comment:
I don't think you need the masterUrl parameter, since you always call it with `client.getMasterUrl().toString()` as argument, you can just do the `client.getMasterUrl().toString()` in this method.

Reply:
I could. But for that I would need to pass the `KubernetesClient` into this method. The `OkHttpClient` does not have the `getMasterUrl()`method. I think passing just the URL as String might be better then passing both clients or converting the client every time it is called.

Reply:
Ah, then fine with the parameter.